flink-tests/src/test/java/org/apache/flink/runtime/operators/lifecycle/validation/TestJobDataFlowValidator.java

```diff
@@ -0,0 +1,147 @@
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/** Check that all data from the upstream reached the respective downstreams. */
+public class TestJobDataFlowValidator {
+    private static final Logger LOG = LoggerFactory.getLogger(TestJobDataFlowValidator.class);
+
+    public static void checkDataFlow(TestJobWithDescription testJob) {
+        Map<String, Map<Integer, OperatorFinishedEvent>> finishEvents = new HashMap<>();
+        for (TestEvent ev : testJob.eventQueue.getAll()) {
+            if (ev instanceof OperatorFinishedEvent) {
+                finishEvents
+                        .computeIfAbsent(ev.operatorId, ign -> new HashMap<>())
+                        .put(ev.subtaskIndex, ((OperatorFinishedEvent) ev));
+            }
+        }
+
+        for (JobVertex upstream : testJob.jobGraph.getVertices()) {
+            for (IntermediateDataSet produced : upstream.getProducedDataSets()) {
+                for (JobEdge edge : produced.getConsumers()) {
+                    Optional<String> upstreamID = getTrackedOperatorID(upstream, true, testJob);
+                    Optional<String> downstreamID =
+                            getTrackedOperatorID(edge.getTarget(), false, testJob);
+                    if (upstreamID.isPresent() && downstreamID.isPresent()) {
+                        checkDataFlow(upstreamID.get(), downstreamID.get(), edge, finishEvents);
+                    } else {
+                        LOG.debug("Ignoring edge (untracked operator): {}", edge);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Check that for each upstream subtask there exists a downstream subtask that received it's
+     * latest emitted element.
+     */
+    private static void checkDataFlow(
+            String upstreamID,
+            String downstreamID,
+            JobEdge edge,
+            Map<String, Map<Integer, OperatorFinishedEvent>> finishEvents) {
+        LOG.debug(
+                "Checking {} edge\n  from {} ({})\n  to {} ({})",
+                edge.getDistributionPattern(),
+                edge.getSource().getProducer().getName(),
+                upstreamID,
+                edge.getTarget().getName(),
+                downstreamID);
+
+        Map<Integer, OperatorFinishedEvent> downstreamFinishInfo =
+                getForOperator(downstreamID, finishEvents);
+
+        Map<Integer, OperatorFinishedEvent> upstreamFinishInfo =
+                getForOperator(upstreamID, finishEvents);
+
+        upstreamFinishInfo.forEach(
+                (upstreamIndex, upstreamInfo) ->
+                        assertTrue(
+                                anySubtaskReceived(
+                                        upstreamID,
+                                        upstreamIndex,
+                                        upstreamInfo.lastSent,
+                                        downstreamFinishInfo.values())));
```

Comment:
can you expand the error message a bit? I had a failure from this error message and I had no idea what has failed.